client/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+
+Note that the max\_age, and no\_cache parameters can be specified by the user in at least three separate locations: (1) in the Cache constructor, (2) in the Cache read method, and (3) in the Transaction read method.  The cache read method determines the most restrictive set of parameters based on these three inputs and sends this most restrictive parameter set to the DB.
+
+Here is how we define the most restrictive parameter set:
+
+* **max\_age**: The most restrictive max\_age is the minimum of all given max\_age values (which we formalize [here](#minRt)).
+
+* **no\_cache**: The most restrictive no\_cache value is the disjunction (boolean "or") of all given no\_cache values.  (If no\_cache == True, the cache must verify with the server by checking the value timestamp it has the most up-to-date value.)
+
+Note that these max\_age and no\_cache values also appear in Cache read method, though users cannot control their values through this method.
+
+### write
+    write(
+        table: string, 
+        op_List: list): value (JSON) throws StaleException  
+
+The write method sends the op\_List to the DB server as a batch write.  
+
+If the server accepts the batch write, the write method also updates the cache to reflect the newly written values.
+
+Otherwise, the write method throws a stale exception, including the value timestamp in the DB that caused the write reject if it is available.
+
+## HTTP Message Formatting
+First, we specify the default and custom headers our HTTP messages utilize.  Then we describe the expected forms of requests/responses between the cache and TreodeDB server.
+
+### Use of HTTP Headers and Directives
+
+In our use of HTTP directives, the term *age* refers to time since the user-specified read\_timestamp.  Specifically, every TreodeDB entry has both a *value timestamp (V_t)* and *read timestamp (R_t)*.  We maintain as an invariant that the given key contained the entry's value between [V\_t, R\_t].
+
+We use the following HTTP directives (under the HTTP header Cache-Control):
+
+* **max-age**: The max number of seconds between our given read\_timestamp and the read\_timestamp of the entry, i.e., (input read\_timestamp) - (entry read\_timestamp) <= max-age.  In other words, we are willing to read a value that could be stale by at most max-age seconds. 
+
+* **no-cache**: Forces cache to send a request to the DB server to retrieve the requested value.  The result cannot just be returned from cache.
+
+An example of usage is below:
+
+    GET /[table_id_string]/[key_string] HTTP/1.1
+    Cache-Control: max-age=3, no-cache
+
+In addition, we use the following custom HTTP header: 
+
+* **Read-TxClock**: Contains the user's specified read\_timestamp.  The value returned by the DB server will be the most recent value older than the specified read\_timestamp value.  Note that Read-TxClock is an long (64 bits).
```

Comment:
We also have Value-TxClock, Condition-TxClock, and Transaction with directives id and item.